parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetWriter.java

```diff
@@ -192,13 +222,17 @@ public Optional<ParquetValueWriters.PrimitiveWriter<?>> visit(
     @Override
     public Optional<ParquetValueWriters.PrimitiveWriter<?>> visit(
         LogicalTypeAnnotation.DateLogicalTypeAnnotation dateType) {
-      return Optional.of(new DateWriter(desc));
+      return Optional.ofNullable(dateWriter(desc));
     }
 
     @Override
     public Optional<ParquetValueWriters.PrimitiveWriter<?>> visit(
         LogicalTypeAnnotation.TimeLogicalTypeAnnotation timeType) {
-      return Optional.of(new TimeWriter(desc));
+      Preconditions.checkArgument(
+          LogicalTypeAnnotation.TimeUnit.MICROS.equals(timeType.getUnit()),
+          "Cannot write time in %s, only MICROS is supported",
+          timeType.getUnit());
+      return Optional.ofNullable(timeWriter(desc));
```

Comment:
Why does this use `ofNullable`? The intent is not to expose `BaseParquetWriter` beyond its package, so there is no need to handle cases where the factory method return `null`. Those are error cases.

Reply:
Because the `InternalWriter` returns null for `timewriter()` meaning use the primitive writer (no need of special handling). Existing `Optional.of()` will give NPE. To avoid it, I used  `Optional.ofNullable()` so it returns `Optional.empty()` when it is null. 

I will do the null check manually and return super method and revert back to `Optional.of()`, if it looks odd.